Tests/GRPCTests/Concurrency/AsyncWriterTests.swift

```diff
@@ -0,0 +1,168 @@
+/*
+ * Copyright 2021, gRPC Authors All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#if compiler(>=5.5)
+@testable import GRPC
+import NIOConcurrencyHelpers
+import XCTest
+
+@available(macOS 12, iOS 15, tvOS 15, watchOS 8, *)
+internal class AsyncWriterTests: GRPCTestCase {
+  func testSingleWriterHappyPath() throws {
+    XCTAsyncTest {
+      let delegate = CollectingDelegate<String, Int>()
+      let writer = AsyncWriter(delegate: delegate)
+
+      try await writer.write("jimmy")
+      XCTAssertEqual(delegate.elements, ["jimmy"])
+
+      try await writer.write("jab")
+      XCTAssertEqual(delegate.elements, ["jimmy", "jab"])
+
+      try await writer.finish(99)
+      XCTAssertEqual(delegate.end, 99)
+    }
+  }
+
+  func testPauseAndResumeWrites() throws {
+    XCTAsyncTest {
+      let delegate = CollectingDelegate<String, Int>()
+      let writer = AsyncWriter(delegate: delegate)
+
+      writer.pause()
+      async let written1: Void = writer.write("wunch")
+      async let written2: Void = writer.write("meat")
+      XCTAssert(delegate.elements.isEmpty)
+
+      writer.resume()
+      try await written1
+      try await written2
+      XCTAssertEqual(delegate.elements.sorted(), ["wunch", "meat"].sorted())
```

Comment:
Why do we have `sorted()` here? I would expect that we check that the elements arrive in order here.